pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -479,63 +394,101 @@ func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
 		return
 	}
 	// remember the layers in the cache as an optimization to avoid searching all remote repositories
-	for _, layer := range manifest.FSLayers {
-		r.cachedLayers.RememberDigest(layer.BlobSum, cacheName)
+	for _, layer := range manifest.References() {
+		r.cachedLayers.RememberDigest(layer.Digest, cacheName)
 	}
 }
 
 // manifestFromImageWithCachedLayers loads the image and then caches any located layers
-func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string) (*schema1.SignedManifest, error) {
-	manifest, err := r.manifestFromImage(image)
+func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string, supportsSchema2 bool) (manifest distribution.Manifest, err error) {
+	// DockerConfigImage is not empty only in manifest 2.
+	isManifest2 := image.DockerConfigImage != ""
+
+	if supportsSchema2 {
+		if isManifest2 {
+			manifest, err = r.deserializedManifestFromImage(image)
+		} else {
+			manifest, err = r.signedManifestFromImage(image)
+		}
+	} else {
+		if isManifest2 {
+			err = fmt.Errorf("Unable to convert new image to old one")
+		} else {
+			manifest, err = r.signedManifestFromImage(image)
+		}
+	}
+
 	if err != nil {
-		return nil, err
+		return
 	}
+
 	r.rememberLayers(manifest, cacheName)
-	return manifest, nil
+	return
 }
 
 // manifestFromImage converts an Image to a SignedManifest.
-func (r *repository) manifestFromImage(image *imageapi.Image) (*schema1.SignedManifest, error) {
+func (r *repository) signedManifestFromImage(image *imageapi.Image) (*schema1.SignedManifest, error) {
+	if image.DockerConfigImage != "" {
+		context.GetLogger(r.ctx).Errorf("old client pulling new image %s", image.DockerImageReference)
+		return nil, fmt.Errorf("Unable to convert new image to old one")
+	}
+
 	dgst, err := digest.ParseDigest(image.Name)
 	if err != nil {
 		return nil, err
 	}
 
 	raw := []byte(image.DockerImageManifest)
-
 	// prefer signatures from the manifest
 	if _, err := libtrust.ParsePrettySignature(raw, "signatures"); err == nil {
-		sm := schema1.SignedManifest{Raw: raw}
+		sm := schema1.SignedManifest{Canonical: raw}
 		if err := json.Unmarshal(raw, &sm); err == nil {
 			return &sm, nil
 		}
 	}
 
+	// We can not use the r.repository here. docker/distribution wraps all the methods that
+	// write or read blobs. It is made for notifications service. We need to get a real
+	// repository without any wrappers.
+	repository, err := Registry.Repository(r.ctx, r.Named())
+	if err != nil {
+		return nil, err
+	}
+
+	manifestService, err := repository.Manifests(r.ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	signaturesGetter, ok := manifestService.(distribution.SignaturesGetter)
+	if !ok {
+		return nil, fmt.Errorf("Unable to convert ManifestService into SignaturesGetter")
+	}
+
 	// Fetch the signatures for the manifest
-	signatures, err := r.Signatures().Get(dgst)
+	signatures, err := signaturesGetter.GetSignatures(r.ctx, dgst)
```

Comment:
I don't see `GetSignatures` being overriden. I believe we need to fetch them from etcd.